source/adapters/opencl/usm.cpp

```diff
@@ -343,18 +404,59 @@ UR_APIEXPORT ur_result_t UR_APICALL urEnqueueUSMFill2D(
     [[maybe_unused]] uint32_t numEventsInWaitList,
     [[maybe_unused]] const ur_event_handle_t *phEventWaitList,
     [[maybe_unused]] ur_event_handle_t *phEvent) {
-  return UR_RESULT_ERROR_INVALID_OPERATION;
+  return UR_RESULT_ERROR_UNSUPPORTED_FEATURE;
 }
 
 UR_APIEXPORT ur_result_t UR_APICALL urEnqueueUSMMemcpy2D(
-    [[maybe_unused]] ur_queue_handle_t hQueue, [[maybe_unused]] bool blocking,
-    [[maybe_unused]] void *pDst, [[maybe_unused]] size_t dstPitch,
-    [[maybe_unused]] const void *pSrc, [[maybe_unused]] size_t srcPitch,
-    [[maybe_unused]] size_t width, [[maybe_unused]] size_t height,
-    [[maybe_unused]] uint32_t numEventsInWaitList,
-    [[maybe_unused]] const ur_event_handle_t *phEventWaitList,
-    [[maybe_unused]] ur_event_handle_t *phEvent) {
-  return UR_RESULT_ERROR_INVALID_OPERATION;
+    ur_queue_handle_t hQueue, bool blocking, void *pDst, size_t dstPitch,
+    const void *pSrc, size_t srcPitch, size_t width, size_t height,
+    uint32_t numEventsInWaitList, const ur_event_handle_t *phEventWaitList,
+    ur_event_handle_t *phEvent) {
+  cl_context CLContext;
+  CL_RETURN_ON_FAILURE(clGetCommandQueueInfo(
+      cl_adapter::cast<cl_command_queue>(hQueue), CL_QUEUE_CONTEXT,
+      sizeof(cl_context), &CLContext, nullptr));
+
+  clEnqueueMemcpyINTEL_fn FuncPtr = nullptr;
+  ur_result_t RetVal = cl_ext::getExtFuncFromContext<clEnqueueMemcpyINTEL_fn>(
+      CLContext, cl_ext::ExtFuncPtrCache->clEnqueueMemcpyINTELCache,
+      cl_ext::EnqueueMemcpyName, &FuncPtr);
+
+  if (!FuncPtr) {
+    return RetVal;
+  }
+
+  std::vector<cl_event> Events;
+  for (size_t HeightIndex = 0; HeightIndex < height; HeightIndex++) {
+    cl_event Event = nullptr;
+    auto ClResult =
+        FuncPtr(cl_adapter::cast<cl_command_queue>(hQueue), false,
+                static_cast<uint8_t *>(pDst) + dstPitch * HeightIndex,
+                static_cast<const uint8_t *>(pSrc) + srcPitch * HeightIndex,
+                width, numEventsInWaitList,
+                cl_adapter::cast<const cl_event *>(phEventWaitList), &Event);
+    Events.push_back(Event);
+    if (ClResult != CL_SUCCESS) {
+      for (const auto &E : Events) {
+        clReleaseEvent(E);
+      }
+      CL_RETURN_ON_FAILURE(ClResult);
+    }
+  }
+  cl_int ClResult = CL_SUCCESS;
+  if (blocking) {
+    ClResult = clWaitForEvents(Events.size(), Events.data());
+  }
+  if (phEvent && ClResult == CL_SUCCESS) {
+    ClResult = clEnqueueBarrierWithWaitList(
+        cl_adapter::cast<cl_command_queue>(hQueue), Events.size(),
+        Events.data(), cl_adapter::cast<cl_event *>(phEvent));
+  }
+  for (const auto &E : Events) {
+    clReleaseEvent(E);
```

Comment:
Probably should add CL_RETURN_ON_ERROR to this call as well. Same for line 441

Reply:
Added here, I'm going to leave 441 as it is to prioritise returning the error from the memcpy